### ISSUE TITLE
SeqIO.parse & File.as_handle bug

### ISSUE THREAD

**Issue by mdehoon** [biopython/biopython#2314]
The `parse` function in `Bio.SeqIO` looks like this:
```python
    with as_handle(handle, mode) as fp:
        if format in _FormatToIterator:
            iterator_generator = _FormatToIterator[format]
            if alphabet is None:
                i = iterator_generator(fp)
        ....
        # This imposes some overhead... wait until we drop Python 2.4 to fix it
        for r in i:
            yield r
```

I am wondering what this code should become now that we have dropped Python 2.4. While `i` is (in most cases) a generator function, simply doing `return i` will not work as the file handle gets closed as soon as we leave the `with` block.

We could move the `with` statement to the generator function in each of the submodules of `Bio.SeqIO`. However, the `iterator_generator` may be an iterable class instead of a generator function (for example, `Bio.SeqIO.SeqXmlIO` does this). In such cases, a `with` block would not work.

Unfortunately the `as_handle` function in `Bio.File` is designed to work as a context manager, which means it assumes that the handle returned by `as_handle` will be used in a `with` block. My suggestion is to remove `@contextlib.contextmanager` from `as_handle`, and use a `try ... finally` block in each submodule of `Bio.SeqIO`.

**Comment by mdehoon** [comment]
The `parse` function in `Bio/AlignIO/__init__.py` also uses `as_handle` to start a `with` block. There the suggestion is made to use `yield from i` once we drop python2. However, the advantage of being able to do `return i` is that the iterator may contain additional information that gets lost if we do `yield from i`.

**Comment by peterjc** [comment]
Currently the internal API gives each parser a file handle (in the expected mode), only the top level API allows a file handle or a filename. I don't immediately see any benefit.

(Also having the top level list of binary formats is quite handy for the test suite, but that is minor.)

Is that example correct? I don't see how it could work, unless you return a custom iterator object which _also_ supports dictionary-like access... kind of how ``Bio.SeqIO.index(...)`` or ``Bio.SeqIO.index_db(...)`` work, perhaps with a bit of the lazy-loading we explored on a GSoC project years ago?

If so, with a custom extended iterator object, I begin to see why you would prefer to do ``return i`` over ``yield from i``.

**Comment by mdehoon** [comment]
Fixed by #2321
